feat: flag to toggle global shell settings

rajatagarwal-ibm opened the pull request:
### Description

Flag to toggle global shell settings

### Release required?
<!--- Identify the type of release. For information about the changes in a semantic versioning release, see [Release versioning](https://terraform-ibm-modules.github.io/documentation/#/versioning). --->

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

<!--- If a release is required, replace this text with information that users need to know about the release. Write the release notes to help users understand the changes, and include information about how to update from the previous version.

Your notes help the merger write the commit message for the PR that is published in the release notes for the module. --->

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

rajatagarwal-ibm commented:
/run pipeline

rajatagarwal-ibm commented:
/run pipeline

ocofaigh commented on variables.tf:
```diff
@@ -162,4 +162,9 @@ variable "cbr_zones" {
 
+variable "skip_cloud_shell_calls" {
+  type        = bool
+  description = "Enable/Disable Global Cloud Shell settings in the account."
```
This description is wrong - its not enabling or disabling anything - its skipping the setting all together

ocofaigh commented on main.tf:
```diff
@@ -54,7 +64,7 @@ locals {
   iam_allowed_ip_addresses_control_mode = var.enforce_allowed_ip_addresses == false ? "MONITOR" : "RESTRICT"
   account_public_access                 = ibm_iam_access_group_account_settings.iam_access_group_account_settings.public_access_enabled
-  account_shell_settings_status         = ibm_cloud_shell_account_settings.cloud_shell_account_settings.enabled
+  account_shell_settings_status         = var.skip_cloud_shell_calls == true ? ibm_cloud_shell_account_settings.cloud_shell_account_settings[0].enabled : var.shell_settings_enabled
```
you want this to default to null if the check is skipped so its not included as an output

ocofaigh commented on main.tf:
```diff
@@ -4,12 +4,21 @@
+locals {
+  # If shell_settings_enabled is false, then skip_cloud_shell_calls must be true.
+  cloud_shell_settings_validate_condition = var.shell_settings_enabled == false && var.skip_cloud_shell_calls == false
+  cloud_shell_settings_validate_msg       = "'skip_cloud_shell_calls' must be set to true when 'shell_settings_enabled = true'"
```
I don't understand this? If we skip the cloud shell calls, the value of `shell_settings_enabled` is irrelevant

ocofaigh replied:
Like cloud shell may be enabled or disabled in the account, so we can't assume its going to be enabled if we skip the check

ocofaigh reviewed (changes requested):
I left some comments - Wont we also need to override the value of `skip_cloud_shell_calls` in the tests to `true` for them to pass (at least until the backend issue is resolved)

rajatagarwal-ibm commented:
/run pipeline

rajatagarwal-ibm commented:
Addressed all comments with the following logic:

1. Defaults `skip_cloud_shell_calls` to false.
2. Create data source `ibm_iam_account_settings` only when `skip_cloud_shell_calls` is false.
3. Create resource `ibm_cloud_shell_account_settings` only when `skip_cloud_shell_calls` is false.
4. Output `account_shell_settings_status` defaults to `null` when `skip_cloud_shell_calls` is true, otherwise its `ibm_cloud_shell_account_settings.cloud_shell_account_settings[0].enabled`. 
5. Updated custom example with `skip_cloud_shell_calls` set to false. 
6. As `shell_settings_enabled` is not part of the tests, there is no point adding `skip_cloud_shell_calls` to the tests too.

rajatagarwal-ibm commented:
/run pipeline

ocofaigh commented on main.tf:
```diff
@@ -4,16 +4,24 @@
 
+locals {
+  # If shell_settings_enabled is true, then skip_cloud_shell_calls must be false.
+  cloud_shell_settings_validate_condition = var.shell_settings_enabled == true && var.skip_cloud_shell_calls == true
```
@rajatagarwal-ibm why are we doing this validation at all? If `skip_cloud_shell_calls` is set to true, then it doesn't matter what the value of `shell_settings_enabled` is because you have just told terraform to skip checking and setting cloud shell settings

ocofaigh replied:
perhaps update the variable description of `shell_settings_enabled` to make it clear that if `skip_cloud_shell_calls` is true, the setting is ignored

ocofaigh commented:
/run pipeline

rajatagarwal-ibm commented:
/run pipeline

terraform-ibm-modules-ops commented:
:tada: This PR is included in version 2.10.0 :tada:

The release is available on [GitHub release](https://github.com/terraform-ibm-modules/terraform-ibm-iam-account-settings/releases/tag/v2.10.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
